cpp/src/arrow/filesystem/filesystem.cc

```diff
@@ -738,6 +761,109 @@ Result<std::shared_ptr<FileSystem>> FileSystemFromUriReal(const Uri& uri,
 
 }  // namespace
 
+Status RegisterFileSystemFactory(std::vector<std::string> schemes,
+                                 FileSystem::Factory factory) {
+  auto& [mutex, scheme_to_factory] = FileSystemFactoryRegistry();
+  std::unique_lock lock{mutex};
+
+  for (auto&& scheme : schemes) {
+    if (auto& ref = scheme_to_factory[scheme]) {
+      return Status::KeyError(
+          "Tried to add a factory for ", scheme,
+          ":// uris, but a factory was already registered for that scheme");
+    } else {
+      ref = factory;
+    }
+  }
+  return Status::OK();
+}
+
+// XXX alternative approach:
+// If we distrust global constructors more than we distrust
+// symbol visibility, we could export specially named functions
+// and retrieve them via dlsym.
+Status RegisterFileSystemFactoryModule(std::string module_name) {
+  static auto* handle = dlopen(nullptr, RTLD_NOW | RTLD_LOCAL);
+  if (!handle) {
+    return Status::Invalid("dlopen failed: ", dlerror());
+  }
+
+  dlerror();
+
+  module_name = "ArrowFileSystemModule_" + module_name;
+  auto* module = reinterpret_cast<void (*)(Status*)>(dlsym(handle, module_name.c_str()));
+  if (!module) {
+    return Status::Invalid("dlsym failed: ", dlerror());
+  }
+
+  Status status;
+  module(&status);
+  return status;
+}
+
+// Does not result in a visible symbol on my machine:
+extern "C" {
+ARROW_EXPORT void ArrowFileSystemModule_slow(Status* status) {
+  *status = RegisterFileSystemFactory(
+      {"slow+file"},
+      [](const Uri& uri, const io::IOContext& io_context,
+         std::string* out_path) -> Result<std::shared_ptr<FileSystem>> {
+        auto local_uri = uri.ToString().substr(sizeof("slow+") - 1);
+        ARROW_ASSIGN_OR_RAISE(auto base_fs,
+                              FileSystemFromUri(local_uri, io_context, out_path));
+        double average_latency = 1;
+        int32_t seed = 0xDEADBEEF;
+        ARROW_ASSIGN_OR_RAISE(auto params, uri.query_items());
+        for (const auto& [key, value] : params) {
+          if (key == "average_latency") {
+            average_latency = std::stod(value);
+          }
+          if (key == "seed") {
+            seed = std::stoi(value, nullptr, /*base=*/16);
+          }
+        }
+        return std::make_shared<SlowFileSystem>(base_fs, average_latency, seed);
+      });
+}
+}
+
+struct Registrar {
+  Registrar(std::vector<std::string> schemes, FileSystem::Factory factory) {
+    Status st = RegisterFileSystemFactory(std::move(schemes), std::move(factory));
+    if (!st.ok()) {
+      st.Abort();
+    }
+  }
+};
+
+// Global constructors will finish execution before main() starts or before
+// dlopen()/LoadLibrary() returns if dynamically loaded.
+Registrar kSlowFileSystemModule{
```

Comment:
This should only be done in filesystem tests.

Reply:
Agreed, I mostly wanted the examples of the two approaches to be close together for the draft review